File: leetcode-practice-kotlin/src/main/kotlin/io/hank/leetcode/kotlin/practices/_005_Longest_Palindromic_Substring.kt
package io.hank.leetcode.kotlin.practices

/**
 * Given a string s, return the longest palindromic substring in s.
 *
 * Example 1:
 *
 * Input: s = "babad"
 *
 * Output: "bab"
 *
 * Explanation: "aba" is also a valid answer.
 *
 * Example 2:
 *
 * Input: s = "cbbd"
 *
 * Output: "bb"
 *
 * Constraints:
 *
 * 1 <= s.length <= 1000
 *
 * s consist of only digits and English letters.
 */
class _005_Longest_Palindromic_Substring {
    fun longestPalindrome(s: String): String {
        if (s.length < 2) {
            return s
        }

        val processed: String = StringBuilder().also {
            it.append("#")
            s.forEach { c ->
                it.append(c).append("#")
            }
        }.toString()
        val n = processed.length
        var maxLength = 1
        var maxCenter = 0
        var center = 0
        var right = 0
        val dp = IntArray(n)

        for (i in 0..<n) {
            // Manacher's Algorithm
            if (i < right) {
                dp[i] = dp[2 * center - i].coerceAtMost(right - i)
            }

            while (dp[i] < i && i + dp[i] < n - 1 && processed[i - dp[i] - 1] == processed[i + dp[i] + 1]) {
                dp[i]++
            }

            // move center and right
            (center + dp[i]).takeIf { it > right }?.let {
                center = i
                right = it
            }

            dp[i].takeIf { it > maxLength }?.let {
                maxLength = it
                maxCenter = i
            }
        }

        return ((maxCenter - maxLength) / 2).run {
            s.substring(this, this + maxLength)
        }
    }
}

File: leetcode-practice-typescript/src/_005_Longest_Palindromic_Substring.ts
/**
 * Given a string s, return the longest palindromic substring in s.
 * <p>
 * Example 1:
 * <p>
 * Input: s = "babad"
 * <p>
 * Output: "bab"
 * <p>
 * Explanation: "aba" is also a valid answer.
 * <p>
 * Example 2:
 * <p>
 * Input: s = "cbbd"
 * <p>
 * Output: "bb"
 * <p>
 * Constraints:
 * <p>
 * 1 <= s.length <= 1000
 * <p>
 * s consist of only digits and English letters.
 */
function longestPalindrome(s: string): string {
    if (s.length < 2) {
        return s;
    }

    const sb = ['#'];
    for (const c of s) {
        sb.push(c);
        sb.push('#');
    }
    const processed = sb.join('');

    const n = processed.length;
    let maxCenter = 0;
    let maxLength = 0;
    let center = 0;
    let right = 0;
    const dp = new Array(n).fill(0);

    for (let i = 0; i < n; i++) {
        if (i < right) {
            dp[i] = Math.min(dp[2 * center - i], right - i);
        }

        while (
            dp[i] < i &&
            i + dp[i] < n - 1 &&
            processed.charAt(i - dp[i] - 1) === processed.charAt(i + dp[i] + 1)
        ) {
            dp[i]++;
        }

        if (i + dp[i] > right) {
            center = i;
            right = i + dp[i];
        }

        if (dp[i] > maxLength) {
            maxLength = dp[i];
            maxCenter = i;
        }
    }

    const realCenter = (maxCenter - maxLength) / 2;
    return s.substring(realCenter, realCenter + maxLength);
}

export function execute() {
    console.log('==== Problem: _005_Longest_Palindromic_Substring ====');

    // Example 1
    let s = 'babad';
    console.log(`Input: s = ${s}`);
    console.log(`Output: ${longestPalindrome(s)}`);

    // Example 2
    s = 'ss';
    console.log(`Input: s = ${s}`);
    console.log(`Output: ${longestPalindrome(s)}`);

    // Example 3
    s = 'cbbd';
    console.log(`Input: s = ${s}`);
    console.log(`Output: ${longestPalindrome(s)}`);
}
